reana_cluster/backends/kubernetes/k8s.py

```diff
@@ -0,0 +1,510 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2017 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation; either version 2 of the License, or (at your option) any later
+# version.
+#
+# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
+# Suite 330, Boston, MA 02111-1307, USA.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization or
+# submit itself to any jurisdiction.
+"""Abstract Base Class representing REANA cluster backend."""
+
+import pkg_resources
+from kubernetes.client.rest import ApiException
+
+import json
+import logging
+import yaml
+
+from pkg_resources import parse_version
+
+from kubernetes import client as k8s_client, config as k8s_config
+from kubernetes.client import configuration
+
+from jinja2 import Environment, FileSystemLoader, TemplateNotFound
+
+from reana_cluster import ReanaBackendABC
+
+
+class KubernetesBackend(ReanaBackendABC):
+    # TODO: Docstring
+
+    __cluster_type = 'kubernetes'
+
+    _conf = {
+        'templates_folder': pkg_resources.resource_filename(
+            __name__, '/templates'),
+        'min_version': 'v1.6.4',
+        'max_version': 'v1.6.4',
+    }
+
+    def __init__(self,
+                 cluster_spec,
+                 cluster_conf=None,
+                 kubeconfig=None,
+                 context=None):
+
+        # Load Kubernetes cluster configuration. Defaults to '~/.kube/config'
+        k8s_config.load_kube_config(kubeconfig, context)
+
+        # Instantiate clients for various Kubernetes REST APIs
+        self.corev1api = k8s_client.CoreV1Api()
+        self.versionapi = k8s_client.VersionApi()
+        self.extbetav1api = k8s_client.ExtensionsV1beta1Api()
+
+        self.cluster_spec = cluster_spec
+        self.cluster_conf = self.generate_configuration(cluster_spec)
+
+    @property
+    def cluster_type(self):
+        return self.__cluster_type
+
+    @property
+    def cluster_type(self):
+        return self.cluster_conf
+
+    @property
+    def current_config(self):
```

Comment:
Since this is `k8s` specific config the naming might be a bit misleading (i.e. It can be confused with the previous one `cluster_config`) Since we call the class `KubernetesBackend` something like `backend_config` would be less confusing, anyhow, we have to fix this naming issue ASAP let see with @tiborsimko in todays stand up.

Reply:
`backend_config`, is perfect for now. Thanks